Fix blending for raw_space modules

jenshannoschwalm opened the pull request:
We have to take care of channels while allocating temporary buffers in OpenCL blending code.

Fixes #14843 

For master and possibly a 4.4.2

jenshannoschwalm commented:
Maybe for release notes

Fix OpenCL mask blending for pre-demosaic modules

TurboGit commented:
Don't seem to be fixed. What I see:

- When I open the RAW in darkroom and after selecting reconstruct LCh

![image](https://github.com/darktable-org/darktable/assets/467069/61a6c1ad-e431-47ab-ab79-07db1c2243f8)

- When I deactivate the module:

![image](https://github.com/darktable-org/darktable/assets/467069/1969a505-2c0f-46c0-b4d8-1f56aeb57b2d)

- So it indeed does something (recovering the pink highlight), now let's add a shape to recover only part of the image:

![image](https://github.com/darktable-org/darktable/assets/467069/d16fae12-1200-4349-9528-431d73984454)

The recovering seems to be deactivated for the whole image.

I'm using OpenCL but I have the same behavior with CPU path.

jenshannoschwalm commented:
Ok, can reproduce (sometimes). If you look at the log with `-d pipe` and possible `-d opencl` if you are using that you see the "work area mismatch" in the log so there was no blending. Investigating if this
a) is something that should be fixed in blending code or
b) there is something wrong in hlr module.

This pr is correct so far ...

jenshannoschwalm commented:
@TurboGit and others, this has become at at least a three-part "fixing story" on blending modules in raw-space.

1. The first commit made sure we have correct dimensions for temp cl buffers so avoided a fallback to cpu code
2. While testing the masking on opposed mode i was surprised that seemed to work in most cases, sometimes it failed. I took some time to understand this, the "symptom" was, the blending failed both in cpu and CL code because of "skip blending, work area mismatch" errors. Underlying reason was a wrong `roi_in` modification in modules expanding to full-area (highlights using opposed/segmentation or the raw cacorrect module). In both cases we must not use full-area provided by the sensor but cropped-by-rawprepare. 
3. There is one problem not understood yet.
     a) Mask blending in laplacian seems never working
     I am not sure if/how blending in non-demosaiced data works.

jenshannoschwalm commented:
Release note would be:

Fixed wrong allocation of opencl image buffers for blending in DEVELOP_BLEND_CS_RAW
Fixed roi_in calculation in highlights and cacorrect modules

TurboGit commented:
> 3\. I can't reproduce problem as commented by @TurboGit , at least not on images/modules i have in use. Could you provide a "failing" image/xmp?

I've just sent you a link to the RAW + XMP. Do not hesitate to ask if anything not clear. TIA.

TurboGit commented:
@jenshannoschwalm : BTW, do you want me to merge this PR now or do you prefer waiting for the blending fix?

jenshannoschwalm commented:
I think it's ready to be merged.

jenshannoschwalm commented:
Will have a look for your file after merging, ok?

jenshannoschwalm commented:
Ok, good point - i can reproduce with your sidecar.
Bad point, i can't generate such a condition myself, don't understand the reason. I can
1. start with a fresh history -> no issue
2. go back in your xmp history to highlights without a mask, generate one and --> no issue
3. I can create another mask in highlights, the issue persists in your mask and no issue in the new mask.

Do you have a clue? Maybe keep un-merged until issue understood and fixed ...

TurboGit commented:
No idea, I tried with and without OpenCL and I can reproduce the issue.
I have tried with CR2 & non CR2 and can reproduce the issue.
I have tried with portrait and landscape and can reproduce the issue.

Really weird that you cannot reproduce... No clue why at the moment.

Note: I have a power outage planned tomorrow for 8h, so I may not be able to reply very quickly.

jenshannoschwalm commented:
No idea either, until we have this fixed i removed the roi_in commit and did one for debugging. I can confirm that the mask_form seems to be found and at least is reported to be rendered correctly.

If i keep your mask - i can change position, size ... - it won't have effect. If i add a second one like shown here, no problem with "getting used". 

Aah, if i modify mask opacity to the full right there is some effect with your mask.

![Bildschirmfoto vom 2023-07-10 07-10-41](https://github.com/darktable-org/darktable/assets/50982232/68610ad4-e906-4443-89bb-d26ef9adcd7e)

TurboGit commented:
@jenshannoschwalm : Stupid me.... My mask was having an opacity of 5% ! I didn't thought about this, sorry for the noise. It works fine now.

TurboGit commented:
EDIT: Our messages crossed, and we came to the same conclusion, a procedural mistake on my side :)

jenshannoschwalm commented:
Maybe we can find a way to better visualize the mask opacity in the module using it?

Anyway, i think this could be merged now, either with the debug stuff included or just the first commit - as you like.

The roi_in stuff was wrong. We need some other way to get this properly, maybe via some rawprepare work ... don't know yet.

TurboGit commented:
I'll merge both commits on master and only first one on 4.4.x branch.

TurboGit reviewed (approved):
Thanks!
